go/util/receipts/receipts.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package receipts
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/hash"
+	"golang.org/x/crypto/nacl/secretbox"
+)
+
+// Data stores parsed receipt data.
+type Data struct {
+	Database string
+	Date     time.Time
+}
+
+// KeySize is the size in bytes of receipt keys.
+const KeySize = 32 // secretbox
+
+// Key is used to encrypt receipt data.
+type Key [KeySize]byte
+
+// Don't use a nonce when sealing/opening secretbox because these receipts need
+// to be used across sessions.
+var emptyNonce [24]byte
+
+// Force all receipts to have the same size, to avoid size attacks.
+const receiptSize = 256
+
+// DecodeKey converts a base64 encoded string to a receipt key.
+func DecodeKey(s string) (key Key, err error) {
+	keySlice, err2 := base64.URLEncoding.DecodeString(s)
+	if err2 != nil {
+		err = err2
+		return
+	}
+
+	if len(keySlice) != len(key) {
+		err = fmt.Errorf("--key must be %d bytes when decoded, not %d", len(key), len(keySlice))
+		return
+	}
+
+	copy(key[:], keySlice)
+	return
+}
+
+// Generate returns a receipt for Data, which is an encrypted query string
+// encoded as base64.
+func Generate(key Key, data Data) (string, error) {
+	d.PanicIfTrue(data.Database == "" || data.Date == (time.Time{}))
+
+	receiptPlainOrig := []byte(url.Values{
```

Comment:
Why use querystring as the encoding? it seems somewhat wasteful to repeat the key names in every receipt. Why not using more like a straight binary encoding ... assuming that first thing is hash of db name, second thing date, etc?